libraries/chain/webassembly/binaryen.cpp

```diff
@@ -0,0 +1,229 @@
+#include <eosio/chain/webassembly/binaryen.hpp>
+#include <eosio/chain/apply_context.hpp>
+
+#include <wasm-binary.h>
+
+
+namespace eosio { namespace chain { namespace webassembly { namespace binaryen {
+
+void entry::call(const string &entry_point, LiteralList& args, apply_context &context)
+{
+   interpreter_interface local_interface(memory, table, import_lut, sbrk_bytes);
+   interface = &local_interface;
+   interpreter_instance local_instance(*module.get(), interface);
+   instance = &local_instance;
+   local_instance.callExport(Name(entry_point), args);
+}
+
+void entry::call_apply(apply_context& context)
+{
+   LiteralList args = {Literal(uint64_t(context.act.account)),
+                       Literal(uint64_t(context.act.name))};
+
+   call("apply", args, context);
+}
+
+void entry::call_error(apply_context& context)
+{
+   LiteralList args = {/* */};
+   call("error", args, context);
+}
+
+int entry::sbrk(int num_bytes) {
+   // TODO: omitted checktime function from previous version of sbrk, may need to be put back in at some point
+   constexpr uint32_t NBPPL2  = 16;
+   const uint32_t         num_pages      = module->memory.initial;
+   const uint32_t         min_bytes      = (num_pages << NBPPL2) > UINT32_MAX ? UINT32_MAX : num_pages << NBPPL2;
+   const uint32_t         prev_num_bytes = sbrk_bytes; //_num_bytes;
+
+   // round the absolute value of num_bytes to an alignment boundary
+   num_bytes = (num_bytes + 7) & ~7;
+
+   if ((num_bytes > 0) && (prev_num_bytes > (wasm_constraints::maximum_linear_memory - num_bytes)))  // test if allocating too much memory (overflowed)
+      throw eosio::chain::page_memory_error();
+   else if ((num_bytes < 0) && (prev_num_bytes < (min_bytes - num_bytes))) // test for underflow
+      throw eosio::chain::page_memory_error();
+
+   // update the number of bytes allocated, and compute the number of pages needed
+   sbrk_bytes += num_bytes;
+   auto num_accessible_pages = (sbrk_bytes + Memory::kPageSize - 1) >> NBPPL2;
+   instance->set_accessible_pages(num_accessible_pages);
+   return prev_num_bytes;
+}
+
+void entry::reset(const info& base_info) {
+   const auto& image = base_info.mem_image;
+   char *base = memory.data;
+   memcpy(base, image.data(), image.size());
+   if (sbrk_bytes > image.size()) {
+      memset(memory.data + image.size(), 0, sbrk_bytes - image.size());
+   }
+   sbrk_bytes = base_info.default_sbrk_bytes;
+   interface = nullptr;
+}
+
+void entry::prepare( const info& base_info) {
+}
+
+static Expression* create_call_checktime(Import *checktime, Module &module) {
+   // create a calling expression
+   CallImport *call = module.allocator.alloc<CallImport>();
+   Const *value = module.allocator.alloc<Const>();
+   value->finalize();
+
+   value->set(Literal(11));
```

Comment:
At some point, we want to change this to 0 (or some invalid value??), I had put 11 just as a place holder if the code to accumulate the count was added to apply_context.